### ISSUE TITLE
Feature Request: Word Addressable Memory Domains

### ISSUE THREAD

**Issue by alyosha-tas** [TASVideos/BizHawk#794]
Intellivision has word addressable memory. Each address cooresponds to a 16 bit value.

It would be nice to be able to correctly represent this in things like the Hex Editor but I don't currently see a way to do it.

I'm not sure if Intellivision is the only system with this issue, but I'm putting this here so it's not forgotten that it's a problem in case the issue comes up later.

**Comment by zeromus** [comment]
it's a problem in other cores that's just ignored. snes vram and psx spu memories are word addressible. heck, you only address the psx spu with u64 units. i'm sure there's some vram-likes for other systems (maybe genesis) that are word addressible. some roms may be 16- or 32- addressible. we simply don't solve it: we treat it as if it was byte addressible.  This kind of just a fact of life. You see the game write to $30, you remember it means what looks like $60 to bizhawk.

I'm not sure we're representing it _incorrectly_ now. You're wanting to simulate the cpu's view of the memory. We view the memory directly.

FWIW I think it would not be extremely hard to fix hex editor to do this, but I would actually rather NOT--there are other tools which would have to follow it. I wouldnt stop someone from adding that option solely to the hex editor though.

**Comment by adelikat** [comment]
When the hex editor is in 2 byte mode is it not identical to what you want?

We have the ability to set the default byte size of memory domains.  Only the Ram Search tool currently listens to that value though, I think.

**Comment by zeromus** [comment]
i think he would want a row to look like this

```
0    1    2    3    4    5    6    7
AAAA BBBB CCCC DDDD EEEE FFFF 0000 1111
```

There are even deeper consequences, potentially. Depending on how ideally this works. For instance, lua's readbyte() on that domain would be INVALID. readword would be required, and readword(1) would return BBBB.

But like I said, all that's modeling one point of view. The simplified point of view we model is great in its own way.

**Comment by adelikat** [comment]
Okay, I see now.  That makes sense honestly.  

But that's an intimidating task to build and play nice with what we have.  Yikes.

**Comment by YoshiRulz** [comment]
Oh, well halving and doubling addresses in the right place is easy apart from in Hex Editor. I'll try it once Asnivor's done with GDI+.

**Comment by alyosha-tas** [comment]
Yeah, this isn't a request for simple trickery with getting hex Editor to look correct, it's about having actual correctly addressable domains of the right word size.

I wouldn't complain if this was closed for being infeasible or not worth the considerable effort it would take. It seemed like a nice-to-have when I was working on intellihawk at the time.

**Comment by Asnivor** [comment]
I'm a little confused. The abstract `MemoryDomain` class already has a `WordSize` property. Surely the cores just need to set this correctly, and the tools (that require it) be updated to actually respect the supplied wordsize?

**Comment by alyosha-tas** [comment]
Yes. But that's a pretty big 'just needs to do this' in there.

**Comment by YoshiRulz** [comment]
Idea: `MemoryDomain SystemBusNativeAddressing` in `IMemoryDomains`. You could easily have an adapter class implement peek and poke with a scale factor. Tool support would still be the hard part.
